xprocess.py

```diff
@@ -25,7 +25,24 @@ def __init__(self, path, name):
         self.pidpath = self.controldir.join("xprocess.PID")
         self.pid = int(self.pidpath.read()) if self.pidpath.check() else None
 
-    def terminate(self):
+    def _terminate_children(self, proc, timeout=10):
+        """Recursively terminates children of proc but not proc itself.
+
+        This is the default behavior unless explicitly disabled by
+        setting ``kill_proc_tree`` flag to false when calling
+        ``XProcessInfo.terminate``.
+
+        :param proc: The proc whose children will be terminated
+        :return: None
+        """
+        children = proc.children(recursive=True)
+        for p in children:
+            p.terminate()
+        _, alive = psutil.wait_procs(children, timeout=timeout)
+        for p in alive:
+            p.kill()
+
+    def terminate(self, kill_proc_tree=True):
```

Comment:
Let's make this a keyword only argument:

```python
def terminate(self, **kwargs):
    kill_proc_tree = kwargs.pop("kill_proc_tree", True)
    if kwargs:
        raise TypeError("unknown keyword arguments: {}".format(kwargs.keys()))      
    ...
```

This of course can be made nicer once we drop Python 2.7:
```python
def terminate(self, *, kill_proc_tree=True):
    ...
```

Reply:
Done. Since you mentioned python 2.7, I was considering opening an issue to discuss the possibility of dropping it. I would gladly start working on a PR for that if someone give me a green light (or is there a reason to keep it for the time being that I am not seeing?).

Reply:
> Done. Since you mentioned python 2.7, I was considering opening an issue to discuss the possibility of dropping it. I would gladly start working on a PR for that if someone gives me the green light (or is there a reason to keep it for the time being that I am not seeing?).

We usually wait until supporting 2.7 becomes a burden or gets into the way of implementing new functionality. Until that time comes and the maintenance burden is reasonable, I think it is best to keep it working. But to be clear: if there's some new functionality that can only be implemented in Python 3+, I'm glad to use that as an excuse to drop 2.7. 👍 

> Finally, I was tinkering around with tox and had to recreate my venv for testing. I forgot to reinstall pre-commit hooks and only noticed after pushing changes (silly me), so the commit history is a bit polluted. I can rebase this if necessary.

Thanks, but don't worry, we've been squash+merging your latest PRs as their changes are self contained in my opinion. 👍

Reply:
It makes sense, thanks for explaining.